src/controller.c

```diff
@@ -127,6 +130,94 @@ struct rspamd_controller_session {
 	gboolean is_spam;
 };
 
+static gboolean
+rspamd_is_encrypted_password (const gchar * password)
+{
+        const gchar *start, *end;
+        gint64 id;
+        gsize size;
+        gboolean ret = TRUE;
+        
+        if (g_str_has_prefix (password, "$")) {
+                /* Parse id */
+                start = password + 1;
+                end = start;
+                size = 0;
+                while (g_ascii_isdigit (*end)) { 
+                        size++;
+                        end++;
+                }
+                if (size) {
+                        id = g_ascii_strtoll (g_strndup (start, size), (char **) NULL, 10);
+                        if (!id) {
+                                ret = FALSE;
+                        }
+                } else {
+                       ret = FALSE;
+                }
+        } else {
+                ret = FALSE;
+        }
+                
+        return ret;
+}
+
+static const gchar *
+rspamd_encrypted_password_get_str (const gchar * password, gsize skip, gsize * length)
+{
+        const gchar *str, *start, *end;
+        gsize size;
+        
+        start = password + skip;   
+        end = start;
+        size = 0;
+        while (g_ascii_isdigit (*end) || g_ascii_isalpha (*end)) { 
+             size++;
+             end++;
+        }
+        if (size) {
+                str = g_strndup (start, size);
+                *length = size;
+                
+        } else {
+                str = NULL;
+        }
+        
+        return str;
+}
+
+static gboolean
+rspamd_check_encrypted_password (const gchar * password, const gchar * check)
+{
+        const gchar *salt, *hash;
+        gsize salt_len, hash_len;
+	guint8 *key = NULL;
+        gboolean ret = TRUE;
+        
+        /* get salt */
+        salt = rspamd_encrypted_password_get_str (check, 3, &salt_len);
+        /* get hash */
+        hash = rspamd_encrypted_password_get_str (check, 3 + salt_len + 1, &hash_len);
+        /* decode salt */
+        salt = rspamd_decode_base32 (salt, salt_len, &salt_len);
```

Comment:
Again, if you allocate `salt` previously then here you'd loose pointer to it and this would lead to a memory leak. If you will modify your `get_str` routine, then this won't happen